peridynamics/integrators.py

```diff
@@ -18,46 +64,522 @@ def __call__(self):
         This method should be implemennted in every concrete integrator.
         """
 
+    @abstractmethod
+    def _build_special(self):
+        """
+        Build OpenCL kernels special to the chosen integrator.
+
+        This method should be implemented in every concrete integrator.
+        """
+
+    @abstractmethod
+    def _set_special_buffers(self):
+        """
+        Set buffers that are special to the chosen integrator.
+
+        This method should be implemented in every concrete integrator.
+        """
+
+    def build(
+            self, nnodes, degrees_freedom, max_neighbours, coords,
+            volume, family, bc_types, bc_values, force_bc_types,
+            force_bc_values, stiffness_corrections, bond_types, densities):
+        """
+        Build OpenCL programs.
+
+        Builds the programs that are common to all integrators and the
+        buffers which are independent of :class:`Model`.simulation parameters.
+        """
+        self.nnodes = nnodes
+        self.degrees_freedom = degrees_freedom
+        self.max_neighbours = max_neighbours
+        self.densities = densities
```

Comment:
It is not possible to initate these variables with init, since the user will instantiate the integrator variable before the model has been instantiated, so those variables don't exist yet.
Further to this, the integrator class cannot be instantiated after the mode class l has been instantiated since the integrator is required by model.
If it is not a hard requirement then I'm tempted to leave it as it is, since there is a tension between polymorphism, reducing the overhead of Model.simulate, and ease of use on the user end which seems to be in good balance as things are.